frappe/core/doctype/test/test.py

```diff
@@ -1,43 +1,74 @@
 # Copyright (c) 2021, Frappe Technologies and contributors
 # License: MIT. See LICENSE
 
+""" This is a virtual doctype controller for test/demo purposes.
+
+- It uses a JSON file on disk as "backend".
+- All docs are stored in "docs" key of JSON file.
+"""
 import json
+import os
+from typing import Any, TypedDict
 
-# import frappe
+import frappe
 from frappe.model.document import Document
 
+DATA_FILE = "data_file.json"
+
+
+class Data(TypedDict):
+	docs: dict[str, dict[str, Any]]
+
+
+def get_current_data() -> Data:
+	"""Read data from disk"""
+	if not os.path.exists(DATA_FILE):
+		return {"docs": {}}
+
+	with open(DATA_FILE) as f:
+		return json.load(f)
+
+
+def update_data(data: Data) -> None:
+	"""Flush updated data to disk"""
+	with open(DATA_FILE, "w+") as data_file:
```

Comment:
this should be pushed into `os.path.join(frappe.local.site, DATA_FILE)` instead of the base sites folder?